File: modules/sdk/src/main/scala/nakka/sdk/EventSourcedEntity.scala
package nakka.sdk

import nakka.core.*
import nakka.core.effect.*

import scala.collection.mutable

/**
 * An entity whose state is derived by replaying the events it has persisted.
 *
 * Cluster sharding guarantees a single instance per entity id across the whole cluster and delivers
 * one command at a time to it, which is what makes the mutable `currentState` below safe — and what
 * lets handlers be written as ordinary sequential code with no locks and no concurrency reasoning.
 *
 * State can only change by persisting an event. There is no `updateState`.
 */
abstract class EventSourcedEntity[S, E]:

  /**
   * Handler return types, so signatures read `def addItem(i: LineItem): Effect[Done]` rather than
   * repeating the state and event types on every method.
   */
  final type Effect[R]         = EventSourcedEffect[S, E, R]
  final type ReadOnlyEffect[R] = nakka.core.effect.ReadOnlyEffect[S, E, R]

  private var stateOpt: Option[S]                = None
  private var contextOpt: Option[CommandContext] = None

  /** State before any event has been persisted. Must never be null. */
  def emptyState: S

  /**
   * Folds one event into the current state. The single place state changes.
   *
   * Called on replay as well as on new events, so it must be pure: no I/O, no clock reads, no
   * random values — anything else makes a recovered entity diverge from the one that wrote the
   * journal.
   */
  def applyEvent(event: E): S

  /** State as of now: after replay, and after any events persisted earlier this command. */
  protected final def currentState: S =
    stateOpt.getOrElse(
      throw IllegalStateException(
        "currentState is only available inside a command handler or applyEvent"
      )
    )

  protected final def commandContext: CommandContext =
    contextOpt.getOrElse(
      throw IllegalStateException("commandContext is only available inside a command handler")
    )

  /** The declarative effect API for this entity. */
  protected final val effects: EventSourcedEffects[S, E] = new EventSourcedEffects[S, E]()

  // ── Runtime hooks ─────────────────────────────────────────────────────────
  private[nakka] def _setState(state: S): Unit                      = stateOpt = Some(state)
  private[nakka] def _setContext(ctx: Option[CommandContext]): Unit = contextOpt = ctx
  private[nakka] def _applyEvent(state: S, event: E): S =
    _setState(state)
    applyEvent(event)

object EventSourcedEntity:

  /**
   * Declares an event sourced entity to the runtime.
   *
   * Extend this in the entity's companion object and register handlers as `val`s:
   * {{{
   * object CounterEntity
   *     extends EventSourcedEntity.Companion[CounterEntity, Counter, CounterEvent](
   *       ComponentId("counter"),
   *       Codecs.serializer[Counter]("counter"),
   *       Codecs.serializer[CounterEvent]("counter-event")
   *     ):
   *   def create(ctx: EventSourcedEntityContext) = new CounterEntity(ctx)
   *
   *   val increase = command("increase")(_.increase)
   *   val get      = query("get")(_.get)
   * }}}
   *
   * `descriptor` is a `def`, not a `val`: a `val` in this base class would be initialised before
   * the subclass's handler `val`s had run, and would see no handlers.
   */
  abstract class Companion[C <: EventSourcedEntity[S, E], S, E](
      val componentId: ComponentId,
      val stateSerializer: Serializer[S],
      val eventSerializer: Serializer[E]
  ):

    private val bindings = mutable.ListBuffer.empty[HandlerBinding[C]]

    /**
     * The state and event serializers are exposed as givens because replying with state and
     * persisting events are the two most common things a handler does — requiring the developer to
     * redeclare instances that this companion already holds would be ceremony with no safety
     * benefit.
     */
    protected final given serializerForState: Serializer[S] = stateSerializer
    protected final given serializerForEvent: Serializer[E] = eventSerializer

    /** Instantiates the entity. Called once per activation, not per command. */
    def create(ctx: EventSourcedEntityContext): C

    /** How often to snapshot. `None` disables snapshotting for this entity. */
    def snapshotEvery: Option[Int] = Some(100)

    /** Registers a handler that takes one argument and may persist events. */
    protected final def command[I, O](name: String)(
        f: C => I => EventSourcedEffect[S, E, O]
    )(using in: Serializer[I], out: Serializer[O]): CommandHandle[C, I, O] =
      add(
        new CommandHandle[C, I, O](
          componentId,
          MethodName(name),
          readOnly = false,
          in,
          out,
          (c, i) => f(c)(i)
        )
      )

    /** Registers a no-argument handler that may persist events. */
    protected final def command[O](name: String)(
        f: C => EventSourcedEffect[S, E, O]
    )(using out: Serializer[O]): NoArgHandle[C, O] =
      add(new NoArgHandle[C, O](componentId, MethodName(name), readOnly = false, out, f))

    /**
     * Registers a read-only handler taking one argument.
     *
     * The `ReadOnlyEffect` bound is the enforcement: a handler that persists cannot be registered
     * as a query, so the runtime's "serve reads from any replica" decision rests on the type system
     * rather than on a naming convention.
     */
    protected final def query[I, O](name: String)(
        f: C => I => ReadOnlyEffect[S, E, O]
    )(using in: Serializer[I], out: Serializer[O]): CommandHandle[C, I, O] =
      add(
        new CommandHandle[C, I, O](
          componentId,
          MethodName(name),
          readOnly = true,
          in,
          out,
          (c, i) => f(c)(i)
        )
      )

    /** Registers a no-argument read-only handler. */
    protected final def query[O](name: String)(
        f: C => ReadOnlyEffect[S, E, O]
    )(using out: Serializer[O]): NoArgHandle[C, O] =
      add(new NoArgHandle[C, O](componentId, MethodName(name), readOnly = true, out, f))

    private def add[H <: HandlerBinding[C]](handle: H): H =
      bindings += handle
      handle

    /** Everything the runtime needs to host this entity. */
    final def descriptor: EventSourcedEntityDescriptor[C, S, E] =
      val seen = bindings.groupBy(_.name).collect {
        case (name, bs) if bs.sizeIs > 1 => s"handler '$name' registered ${bs.size} times"
      }
      if seen.nonEmpty then
        throw IllegalArgumentException(
          seen.mkString(s"invalid entity '$componentId':\n  - ", "\n  - ", "")
        )

      EventSourcedEntityDescriptor(
        componentId,
        stateSerializer,
        eventSerializer,
        create,
        snapshotEvery,
        bindings.map(b => b.name -> b).toMap
      )

/** The registered form of an event sourced entity. */
final case class EventSourcedEntityDescriptor[C <: EventSourcedEntity[S, E], S, E](
    componentId: ComponentId,
    stateSerializer: Serializer[S],
    eventSerializer: Serializer[E],
    create: EventSourcedEntityContext => C,
    snapshotEvery: Option[Int],
    handlers: Map[MethodName, HandlerBinding[C]]
) extends ComponentDescriptor:
  val kind: ComponentKind = ComponentKind.EventSourcedEntity

  private[nakka] def handler(name: MethodName): Option[HandlerBinding[C]] =
    handlers.get(name)

File: sdks/typescript/src/eventSourcedEntity.ts
// An event sourced entity: state derived from a journal of events. The class declares its identity and
// codecs as statics and its handlers in a static table; the instance holds the state for one command.
//
//   export class ShoppingCartEntity extends EventSourcedEntity<ShoppingCart, ShoppingCartEvent> {
//     static readonly componentId = "shopping-cart"
//     static readonly state = jsonCodec(ShoppingCart, "shopping-cart")
//     static readonly events = jsonCodec(ShoppingCartEvent, "shopping-cart-event")
//     static readonly handlers = { addItem: command("add-item", LineItem, Done, (cart: ShoppingCartEntity, item) => cart.addItem(item)) }
//     emptyState() { ... }   applyEvent(state, event) { ... }   addItem(item) { return this.effects.persist(...).thenReply(() => done) }
//   }

import type { Shape } from "./codec.ts"
import type { CommandContext } from "./context.ts"
import type { HandlerTable } from "./handlers.ts"
import type { ComponentClient } from "./client.ts"
import { EventSourcedEffects } from "./effects/eventSourced.ts"

export abstract class EventSourcedEntity<S, E> {
  /** The effect builders: `this.effects.persist(...)`, `.reply(...)`, `.error(...)`. */
  readonly effects: EventSourcedEffects<S, E> = new EventSourcedEffects<S, E>()

  #state: S | undefined
  #entityId: string | undefined
  #context: CommandContext | undefined
  #client: ComponentClient | undefined
  #bound = false

  /** The state the entity has now, inside a handler. */
  get state(): S {
    if (!this.#bound) throw new Error("state is only available inside a command handler")
    return this.#state as S
  }

  /** The id of the instance this object stands for; available in `emptyState` and in handlers. */
  get entityId(): string {
    if (this.#entityId === undefined) throw new Error("entityId is only available once the entity is bound to an instance")
    return this.#entityId
  }

  /** The command's context: component id, sequence number, metadata, the clock. */
  get context(): CommandContext {
    if (!this.#context) throw new Error("context is only available inside a command handler")
    return this.#context
  }

  /** The component client, scoped to the command's trace. */
  get client(): ComponentClient {
    if (!this.#client) throw new Error("client is only available inside a command handler")
    return this.#client
  }

  /** @internal The kind, read off the prototype by the typed client and the timers. */
  get _kind(): "event-sourced" {
    return "event-sourced"
  }

  /** The state of an instance with no events yet. `this.entityId` is available here. */
  abstract emptyState(): S

  /** The fold: the state after `event`, given the state before it. Must not read `this.state`. */
  abstract applyEvent(state: S, event: E): S

  /** @internal Bound by the runtime and the testkit before `emptyState`. */
  _bindInstance(entityId: string): void {
    this.#entityId = entityId
  }

  /** @internal Bound by the runtime and the testkit for the duration of one handler call. */
  _bindCommand(state: S, context: CommandContext, client: ComponentClient): void {
    this.#state = state
    this.#context = context
    this.#client = client
    this.#bound = true
  }

  /** @internal */
  _unbindCommand(): void {
    this.#state = undefined
    this.#context = undefined
    this.#client = undefined
    this.#bound = false
  }
}

/** The statics an event sourced entity class must declare; `register` constrains on this type. */
export interface EventSourcedEntityClass<S = unknown, E = unknown, C extends EventSourcedEntity<S, E> = EventSourcedEntity<S, E>> {
  new (): C
  readonly componentId: string
  readonly state: Shape<S>
  readonly events: Shape<E>
  readonly handlers: HandlerTable<C>
  /** Ask for a snapshot every N events; absent or 0: never. */
  readonly snapshotEvery?: number
}
